src/org/traccar/database/DeviceManager.java

```diff
@@ -169,76 +90,129 @@ public Device getDeviceByPhone(String phone) {
         return devicesByPhone.get(phone);
     }
 
+    @Override
+    public Set<Long> getAllItems() {
+        Set<Long> result = super.getAllItems();
+        if (result.isEmpty()) {
+            try {
+                updateDeviceCache(true);
+            } catch (SQLException e) {
+                Log.warning(e);
+            }
+            result = super.getAllItems();
+        }
+        return result;
+    }
+
     public Collection<Device> getAllDevices() {
-        boolean forceUpdate = devicesById.isEmpty();
+        return getItems(Device.class, getAllItems());
+    }
 
-        try {
-            updateDeviceCache(forceUpdate);
-        } catch (SQLException e) {
-            Log.warning(e);
+    @Override
+    public Set<Long> getUserItems(long userId) {
+        if (Context.getPermissionsManager() != null) {
+            return Context.getPermissionsManager().getDevicePermissions(userId);
+        } else {
+            return new HashSet<>();
         }
-
-        return devicesById.values();
     }
 
-    public Collection<Device> getDevices(long userId) throws SQLException {
-        Collection<Device> devices = new ArrayList<>();
-        for (long id : Context.getPermissionsManager().getDevicePermissions(userId)) {
-            devices.add(devicesById.get(id));
+    @Override
+    public Set<Long> getManagedItems(long userId) {
+        Set<Long> result = new HashSet<>();
+        result.addAll(getUserItems(userId));
+        for (long managedUserId : Context.getUsersManager().getUserItems(userId)) {
+            result.addAll(getUserItems(managedUserId));
         }
-        return devices;
+        return result;
     }
 
-    public Collection<Device> getManagedDevices(long userId) throws SQLException {
-        Collection<Device> devices = new HashSet<>();
-        devices.addAll(getDevices(userId));
-        for (long managedUserId : Context.getUsersManager().getManagedItems(userId)) {
-            devices.addAll(getDevices(managedUserId));
+    private void putUniqueDeviceId(Device device) {
+        if (devicesByUniqueId == null) {
+            devicesByUniqueId = new ConcurrentHashMap<>(getAllItems().size());
         }
-        return devices;
+        devicesByUniqueId.put(device.getUniqueId(), device);
     }
 
-    public void addDevice(Device device) throws SQLException {
-        dataManager.addObject(device);
+    private void putPhone(Device device) {
+        if (devicesByPhone == null) {
+            devicesByPhone = new ConcurrentHashMap<>(getAllItems().size());
+        }
+        devicesByPhone.put(device.getPhone(), device);
+    }
 
-        devicesById.put(device.getId(), device);
-        devicesByUniqueId.put(device.getUniqueId(), device);
+    @Override
+    protected void addNewItem(BaseModel item) {
```

Comment:
Can we use `Device` directly in the arguments here? Same probably applies to other classes as well.